python/openchange/migration/__init__.py

```diff
@@ -0,0 +1,287 @@
+        """
+        if target_version is None:
+            target_version = max(self.migrations[app].keys())
+
+        go_forward = target_version >= self.version(app)
+
+        # 1) Get migrations sorted
+        # 2) Keep only the migrations to apply/unapply
+        # 3) Call these migrations objects with the cursor as parameter
+        if go_forward:
+            migrations = sorted(self.migrations[app].items())
+            migrations = [m for m in migrations if m[0] <= target_version and m[0] > self.version(app)]
+        else:
+            migrations = sorted(self.migrations[app].items(), reverse=True)
+            migrations = [m for m in migrations if m[0] > target_version and m[0] <= self.version(app)]
+
+        for version, migration in migrations:
+            # Migrate inside a transaction
+            cursor = self.db.cursor()
+            if go_forward:
+                migration.apply(cursor)
+            else:
+                migration.unapply(cursor)
+            self.db.commit()
+            cursor.close()
+            if go_forward:
+                self.apply_version(app, version)
+            else:
+                self.unapply_version(app, version)
+
+        return len(migrations) > 0
+
+    def fake_migration(self, app, target_version):
+        """Fake a migration to a specific target version by
+        simulating the apply/unapply migrations.
+
+        :param str app: the "application" to fake migrations
+        :param int target_version: version to go
+        :rtype: bool
+        :returns: True if any migration was faked
+        """
+        if app not in self.migrations:
+            return False
+
+        if target_version is None:
+            target_version = 0
+
+        current_version = self.version(app)
+        if current_version is None:
+            current_version = 0
+
+        if current_version < target_version:
+            ran = range(current_version + 1, target_version + 1)
+            func = self.apply_version
+        elif current_version > target_version:
+            ran = range(target_version + 1, current_version + 1)
+            func = self.unapply_version
+        else:
+            return False
+
+        for ver in ran:
+            if ver in self.migrations[app]:
+                func(app, ver)
+
+        return True
+
+    def applied_migration(self, app, version):
+        """Give back the timestamp for a given version from an app
+
+        :param str app: the application
+        :param int version: the version
+        :rtype datetime:
+        :returns: None if the version does not have been applied
+        """
+        cursor = self.db.cursor()
+        cursor.execute("""SELECT applied FROM {0}
+                          WHERE app = %s AND version = %s""".format(self.table_name),
+                       (app, version))
+        row = cursor.fetchone()
+        if row and row[0]:
+            return row[0]
+
+        return None
```

Comment:
Sorry, I missed that there was a code path in which it actually returned non-None. Disregard my comment, please :)